File: 0045. Jump Game II/Solution0045.py
class Solution0045:
    def jump(self, nums: List[int]) -> int:
        n = len(nums)
        cur, farthest, steps = 0, 0, 0
        for i in range(n-1):
            if i > cur:
                return -1
            farthest = max(farthest, i + nums[i])
            if farthest >= n-1:
                return steps + 1
            if i == cur:
                cur = farthest
                steps += 1
        return steps

File: 0045. Jump Game II/Solution0045.js
/**
 * @param {number[]} nums
 * @return {number}
 */
var jump = function(nums) {
    let steps = 0, cur = 0, longest = 0, len = nums.length;
    for (let i = 0; i < len-1; i++) {
        if (i > cur) return -1;
        longest = Math.max(longest, i + nums[i]);
        if (longest >= len - 1) return steps + 1;
        if (i == cur) {
            cur = longest; steps++;
        }
    }
    return steps;    
};